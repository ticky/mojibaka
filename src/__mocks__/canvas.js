export const prepareCanvasContext = jest.fn((context) => context);
export const canDrawCharacter = jest.fn();
export const getCharacterWidth = jest.fn();
