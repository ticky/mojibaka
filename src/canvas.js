const DRAW_SIZE = 32;

export function prepareCanvasContext(context) {
  if (!context) {
    context = document.createElement('canvas').getContext('2d');
    context.fillStyle = "#000";
    context.font = `${DRAW_SIZE}px "Apple Color Emoji","Segoe UI","Segoe UI Emoji","Segoe UI Symbol",Arial`;
  } else {
    context.clearRect(0, 0, DRAW_SIZE * 2, DRAW_SIZE * 2);
  }
  return context;
}

export function canDrawCharacter(character, context) {
  context = prepareCanvasContext(context);
  context.fillText(character, 0, DRAW_SIZE);

  return context.getImageData(DRAW_SIZE / 2, DRAW_SIZE / 2, 1, 1).data[0] !== 0;
}

export function getCharacterWidth(character, context) {
  context = prepareCanvasContext(context);

  return context.measureText(character).width;
}
