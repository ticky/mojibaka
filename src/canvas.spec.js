/* global describe, it, beforeEach, afterEach, expect, __dirname, jest */
import { Image, Context2d } from 'canvas';
import path from 'path';
import fs from 'fs';
import TEST_CHARS from './__fixtures__/test-characters';

import { prepareCanvasContext, canDrawCharacter, getCharacterWidth } from './canvas';

// import * as __CANVAS from 'canvas';
// import * as __LOCAL_CANVAS from './canvas';

// console.log('canvas.spec', __CANVAS, __LOCAL_CANVAS, __CANVAS === __LOCAL_CANVAS);

// neither 'canvas' nor './canvas' should be mocked,
// but for some reason 'canvas' remains mocked in this
// file, despite mocking not being requested
// bug filed at <https://github.com/facebook/jest/issues/1927>
jest.unmock('canvas');

const FIXTURE_PATH = path.resolve(__dirname, '__fixtures__');

const PLATFORMS = fs.readdirSync(FIXTURE_PATH)
  .filter((filename) => fs.statSync(path.resolve(FIXTURE_PATH, filename)).isDirectory());

function MOCK_AND_CALL_THROUGH_STORING_RETURN_VALUE(method, returnValueArray) {
  return jest.fn(function() {
    const returnValue = method.apply(this, Array.from(arguments));
    returnValueArray.push(returnValue);
    return returnValue;
  });
}

describe('canvas', () => {
  const _realCreateElement = document.createElement;
  const _realGetContext = HTMLCanvasElement.prototype.getContext;
  let _createdElements = [];
  let _createdContexts = [];

  beforeEach(() => {
    document.createElement = MOCK_AND_CALL_THROUGH_STORING_RETURN_VALUE(_realCreateElement, _createdElements);
    window.HTMLCanvasElement.prototype.getContext = MOCK_AND_CALL_THROUGH_STORING_RETURN_VALUE(_realGetContext, _createdContexts);
  });

  afterEach(() => {
    document.createElement = _realCreateElement;
    window.HTMLCanvasElement.prototype.getContext = _realGetContext;
    _createdElements = [];
    _createdContexts = [];
  });

  describe('prepareCanvasContext', () => {
    describe('with an existing canvas context', () => {
      it(`doesn't reconfigure the font and fillStyle`, () => {
        const context = document.createElement('canvas').getContext('2d');

        context.fillStyle = '#ff00ff';
        context.font = '12px Verdana';

        prepareCanvasContext(context);

        expect(context.fillStyle).toMatchSnapshot();
        expect(context.font).toMatchSnapshot();
        expect(document.createElement).toHaveBeenCalledTimes(1);
        expect(document.createElement).toHaveBeenCalledWith('canvas');
      });

      it('clears the canvas', () => {
        const canvas = document.createElement('canvas');
        const context = canvas.getContext('2d');

        context.fillStyle = '#ff00ff';
        context.fillRect(0, 0, 64, 64);

        expect(canvas.toDataURL('image/png')).toMatchSnapshot();

        prepareCanvasContext(context);

        expect(canvas.toDataURL('image/png')).toMatchSnapshot();
        expect(document.createElement).toHaveBeenCalledTimes(1);
        expect(document.createElement).toHaveBeenCalledWith('canvas');
      });
    });

    describe('with no canvas context', () => {
      it(`creates a canvas with font and fillStyle`, () => {
        prepareCanvasContext();

        expect(document.createElement).toHaveBeenCalledTimes(1);
        expect(document.createElement).toHaveBeenCalledWith('canvas');
        expect(_createdElements.length).toBe(1);
        expect(_createdElements[0].toDataURL('image/png')).toMatchSnapshot();

        expect(window.HTMLCanvasElement.prototype.getContext).toHaveBeenCalledTimes(1);
        expect(window.HTMLCanvasElement.prototype.getContext).toHaveBeenCalledWith('2d');
        expect(_createdContexts.length).toBe(1);
        expect(_createdContexts[0].fillStyle).toMatchSnapshot();
        expect(_createdContexts[0].font).toMatchSnapshot();
      });
    });
  });

  describe('canDrawCharacter', () => {
    const _realFillText = Context2d.prototype.fillText;

    describe('in general', () => {
      beforeEach(() => {
        Context2d.prototype.fillText = jest.fn();
      });

      afterEach(() => {
        Context2d.prototype.fillText = _realFillText;
      });

      describe('with an existing canvas context', () => {
        it('calls through to fillText and confirms drawing', () => {
          const context = document.createElement('canvas').getContext('2d');
          expect(canDrawCharacter('■', context)).toMatchSnapshot();
          expect(document.createElement).toHaveBeenCalledTimes(1);
          expect(document.createElement).toHaveBeenCalledWith('canvas');
          expect(Context2d.prototype.fillText).toHaveBeenCalledTimes(1);
          expect(Context2d.prototype.fillText).toHaveBeenCalledWith('■', 0, 32);
        });
      });

      describe('with no canvas context', () => {
        it('calls through to fillText and confirms drawing without creating a context', () => {
          expect(canDrawCharacter('■')).toMatchSnapshot();
          expect(document.createElement).toHaveBeenCalledTimes(1);
          expect(document.createElement).toHaveBeenCalledWith('canvas');
          expect(Context2d.prototype.fillText).toHaveBeenCalledTimes(1);
          expect(Context2d.prototype.fillText).toHaveBeenCalledWith('■', 0, 32);
        });
      });
    });

    PLATFORMS.forEach((platform) => {
      describe(platform, () => {
        beforeEach(() => {
          Context2d.prototype.fillText = jest.fn(function(text) {
            const fileName = text
              .match(/([\uD800-\uDBFF][\uDC00-\uDFFFF]|[\S\s])/g)
              .map((char) => char.codePointAt(0).toString(16))
              .join('-');

            const image = new Image();
            image.src = fs.readFileSync(path.resolve(
              __dirname,
              `__fixtures__/${platform}/${fileName}.png`
            ));

            this.drawImage(image, 0, 0);
          });
        });

        afterEach(() => {
          Context2d.prototype.fillText = _realFillText;
        });

        TEST_CHARS.forEach((char) => {
          it(`shows expected behaviour for ${char}`, () => {
            expect(canDrawCharacter(char)).toMatchSnapshot();
            expect(document.createElement).toHaveBeenCalledTimes(1);
            expect(document.createElement).toHaveBeenCalledWith('canvas');
            expect(Context2d.prototype.fillText).toHaveBeenCalledTimes(1);
            expect(Context2d.prototype.fillText).toHaveBeenCalledWith(char, 0, 32);
          });
        });
      });
    });
  });

  describe('getCharacterWidth', () => {
    const _realMeasureText = Context2d.prototype.measureText;

    beforeEach(() => {
      Context2d.prototype.measureText = jest.fn(() => ({ width: 1234, height: -1234 }));
    });

    afterEach(() => {
      Context2d.prototype.measureText = _realMeasureText;
    });

    describe('with an existing canvas context', () => {
      it('calls through to measureText and returns the width', () => {
        const context = document.createElement('canvas').getContext('2d');
        expect(getCharacterWidth('foo', context)).toMatchSnapshot();
        expect(document.createElement).toHaveBeenCalledTimes(1);
        expect(document.createElement).toHaveBeenCalledWith('canvas');
        expect(Context2d.prototype.measureText).toHaveBeenCalledTimes(1);
        expect(Context2d.prototype.measureText).toHaveBeenCalledWith('foo');
      });
    });

    describe('with no canvas context', () => {
      it('calls through to measureText and returns the width without creating a context', () => {
        expect(getCharacterWidth('foo')).toMatchSnapshot();
        expect(document.createElement).toHaveBeenCalledTimes(1);
        expect(document.createElement).toHaveBeenCalledWith('canvas');
        expect(Context2d.prototype.measureText).toHaveBeenCalledTimes(1);
        expect(Context2d.prototype.measureText).toHaveBeenCalledWith('foo');
      });
    });
  });
});
