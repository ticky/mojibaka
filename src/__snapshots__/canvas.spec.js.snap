exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for üë∏ 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for üë∏üèæ 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for üïµ 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for üòô 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for üôÇ 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for ü§î 1`] = `true`;

exports[`canvas canDrawCharacter iOS 10 - Safari 10.0 shows expected behaviour for ü§£ 1`] = `false`;

exports[`canvas canDrawCharacter in general with an existing canvas context calls through to fillText and confirms drawing 1`] = `false`;

exports[`canvas canDrawCharacter in general with no canvas context calls through to fillText and confirms drawing without creating a context 1`] = `false`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for üë∏ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for üë∏üèæ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for üïµ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for üòô 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for üôÇ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for ü§î 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Chrome 53.0.2785.143 shows expected behaviour for ü§£ 1`] = `false`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for üë∏ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for üë∏üèæ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for üïµ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for üòô 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for üôÇ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for ü§î 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Firefox 49.0 shows expected behaviour for ü§£ 1`] = `false`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for üë∏ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for üë∏üèæ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for üïµ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for üòô 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for üôÇ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for ü§î 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Opera 40.0.2308.81 shows expected behaviour for ü§£ 1`] = `false`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for üë∏ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for üë∏üèæ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for üïµ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for üòô 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for üôÇ 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for ü§î 1`] = `true`;

exports[`canvas canDrawCharacter macOS Sierra - Safari 10.0 shows expected behaviour for ü§£ 1`] = `false`;

exports[`canvas getCharacterWidth with an existing canvas context calls through to measureText and returns the width 1`] = `1234`;

exports[`canvas getCharacterWidth with no canvas context calls through to measureText and returns the width without creating a context 1`] = `1234`;

exports[`canvas prepareCanvasContext with an existing canvas context clears the canvas 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAABWklEQVR4nO3UwQ3AIBDAMK7770xX4Ici7Anyyuy193rYrJnbDcCZ73YAwCnDAjIMC8gwLCDDsIAMwwIyDAvIMCwgw7CADMMCMgwLyDAsIMOwgAzDAjIMC8gwLCDDsIAMwwIyDAvIMCwgw7CADMMCMgwLyDAsIMOwgAzDAjIMC8gwLCDDsIAMwwIyDAvIMCwgw7CADMMCMgwLyDAsIMOwgAzDAjIMC8gwLCDDsIAMwwIyDAvIMCwgw7CADMMCMgwLyDAsIMOwgAzDAjIMC8gwLCDDsIAMwwIyDAvIMCwgw7CADMMCMgwLyDAsIMOwgAzDAjIMCwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB43A+aOwOAiiU5oAAAAABJRU5ErkJggg=="`;

exports[`canvas prepareCanvasContext with an existing canvas context clears the canvas 2`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAAxUlEQVR4nO3BMQEAAADCoPVPbQhfoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOA1v9QAATX68/0AAAAASUVORK5CYII="`;

exports[`canvas prepareCanvasContext with an existing canvas context doesn't reconfigure the font and fillStyle 1`] = `"#ff00ff"`;

exports[`canvas prepareCanvasContext with an existing canvas context doesn't reconfigure the font and fillStyle 2`] = `"12px Verdana"`;

exports[`canvas prepareCanvasContext with no canvas context creates a canvas with font and fillStyle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAAxUlEQVR4nO3BMQEAAADCoPVPbQhfoAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOA1v9QAATX68/0AAAAASUVORK5CYII="`;

exports[`canvas prepareCanvasContext with no canvas context creates a canvas with font and fillStyle 2`] = `"#000000"`;

exports[`canvas prepareCanvasContext with no canvas context creates a canvas with font and fillStyle 3`] = `"32px \"Apple Color Emoji\",\"Segoe UI\",\"Segoe UI Emoji\",\"Segoe UI Symbol\",Arial"`;
